Add note about Laravel snake_case() in text helper

JohntheFish opened the pull request:
Takes a CamelCase string and turns it into snake_case resolving bugs in uncamelcase(). An alternative method is required  because the bugged behaviour of uncamelcase() has too long a legacy and some code may actually depend on it!

Provides a solution for https://github.com/concretecms/concretecms/issues/12238

Tests:

```
echo '<pre><code>';
$th = app()->make('helper/text');
print_r([
            [
                'Source'              => 'CamelCase',
                'uncamelcase'         => $th->uncamelcase('CamelCase'),
                'expected'            => 'camel_case',
                'camelTo_snake_case' => $th->camelTo_snake_case('CamelCase'),
            ],
            [
                'Source'              => 'camelCase',
                'uncamelcase'         => $th->uncamelcase('camelCase'),
                'expected'            => 'camel_case',
                'camelTo_snake_case' => $th->camelTo_snake_case('camelCase'),
            ],
            [
                'Source'              => 'Camelcase',
                'uncamelcase'         => $th->uncamelcase('Camelcase'),
                'expected'            => 'camelcase',
                'camelTo_snake_case' => $th->camelTo_snake_case('Camelcase'),
            ],
            [
                'Source'              => 'already_snake_case',
                'uncamelcase'         => $th->uncamelcase('already_snake_case'),
                'expected'            => 'already_snake_case',
                'camelTo_snake_case' => $th->camelTo_snake_case('already_snake_case'),
            ],
            [
                'Source'              => 'LongerCamelCase',
                'uncamelcase'         => $th->uncamelcase('LongerCamelCase'),
                'expected'            => 'longerCamelCase',
                'camelTo_snake_case' => $th->camelTo_snake_case('LongerCamelCase'),
            ],
            [
                'Source'              => 'LongerCamelcase',
                'uncamelcase'         => $th->uncamelcase('longerCamelCase'),
                'expected'            => 'longerCamelCase',
                'camelTo_snake_case' => $th->camelTo_snake_case('LongerCamelcase'),
            ],
            [
                'Source'              => 'LongerCamelcase',
                'uncamelcase'         => $th->uncamelcase('LongerCamelcase'),
                'expected'            => 'longer_camelcase',
                'camelTo_snake_case' => $th->camelTo_snake_case('LongerCamelcase'),
            ],
            [
                'Source'              => 'MixedCamelAnd_snake_case',
                'uncamelcase'         => $th->uncamelcase('MixedCamelAnd_snake_case'),
                'expected'            => 'mixed_camel_and_snake_case',
                'camelTo_snake_case' => $th->camelTo_snake_case('MixedCamelAnd_snake_case'),
            ],
            [
                'Source'              => 'mixedCamelAnd_snake_case',
                'uncamelcase'         => $th->uncamelcase('mixedCamelAnd_snake_case'),
                'expected'            => 'mixed_camel_and_snake_case',
                'camelTo_snake_case' => $th->camelTo_snake_case('mixedCamelAnd_snake_case'),
            ],
            [
                'Source'              => 'camelTo_snake_case',
                'uncamelcase'         => $th->uncamelcase('camelTo_snake_case'),
                'expected'            => 'camel_to_snake_case',
                'camelTo_snake_case' => $th->camelTo_snake_case('camelTo_snake_case'),
            ],
        ]);

echo '</code></pre>';
```

Test Output
```
Array
(
    [0] => Array
        (
            [Source] => CamelCase
            [uncamelcase] => camel_case
            [expected] => camel_case
            [camelTo_snake_case] => camel_case
        )

    [1] => Array
        (
            [Source] => camelCase
            [uncamelcase] => case
            [expected] => camel_case
            [camelTo_snake_case] => camel_case
        )

    [2] => Array
        (
            [Source] => Camelcase
            [uncamelcase] => camelcase
            [expected] => camelcase
            [camelTo_snake_case] => camelcase
        )

    [3] => Array
        (
            [Source] => already_snake_case
            [uncamelcase] => 
            [expected] => already_snake_case
            [camelTo_snake_case] => already_snake_case
        )

    [4] => Array
        (
            [Source] => LongerCamelCase
            [uncamelcase] => longer_camel_case
            [expected] => longerCamelCase
            [camelTo_snake_case] => longer_camel_case
        )

    [5] => Array
        (
            [Source] => LongerCamelcase
            [uncamelcase] => camel_case
            [expected] => longerCamelCase
            [camelTo_snake_case] => longer_camelcase
        )

    [6] => Array
        (
            [Source] => LongerCamelcase
            [uncamelcase] => longer_camelcase
            [expected] => longer_camelcase
            [camelTo_snake_case] => longer_camelcase
        )

    [7] => Array
        (
            [Source] => MixedCamelAnd_snake_case
            [uncamelcase] => mixed_camel_and_snake_case
            [expected] => mixed_camel_and_snake_case
            [camelTo_snake_case] => mixed_camel_and_snake_case
        )

    [8] => Array
        (
            [Source] => mixedCamelAnd_snake_case
            [uncamelcase] => camel_and_snake_case
            [expected] => mixed_camel_and_snake_case
            [camelTo_snake_case] => mixed_camel_and_snake_case
        )

    [9] => Array
        (
            [Source] => camelTo_snake_case
            [uncamelcase] => to_snake_case
            [expected] => camel_to_snake_case
            [camelTo_snake_case] => camel_to_snake_case
        )

)
```

JohntheFish committed:
Provide camelTo_snake_case() in text helper

Takes a CamelCase string and turns it into snake_case resolving bugs in uncamelcase(). An alternative method is required 
because the bugged behaviour of uncamelcase() has too long a legacy and some code may actually depend on it!

mlocati commented:
What about adding a phpunit test case?

JohntheFish commented:
Is there any documentation on adding tests?

mlocati commented:
> Is there any documentation on adding tests?

We use PHPUnit: you can read its documentation.

Generally speaking, it's enough to add a class under the `tests/tests` directory that extends `Concrete\Tests\TestCase` (or one of its subclasses like `Concrete\TestHelpers\Database\ConcreteDatabaseTestCase`), and add a public method whose name starts with `test`.

In this particular case, we [already have a class that tests the `Text` service class](https://github.com/concretecms/concretecms/blob/9.3.9/tests/tests/Utility/Service/TextTest.php), so I'd simply add tests there.

Since the manual tests [you wrote](https://github.com/concretecms/concretecms/pull/12410#issue-2782040267) are repeatable, I'd use a dataprovider, which provides all the cases you want to test.

Something like this:

```php
public function camelToSnakeCaseProvider(): array
{
    return [
        ['CamelCase', 'camel_case'],
        ['camelCase', 'camel_case'],
        ['Camelcase', 'camelcase'],
        ['already_snake_case', 'already_snake_case'],
        ['LongerCamelCase', 'longer_camel_case'],
        ['LongerCamelcase', 'longer_camelcase'],
        ['longerCamelCase', 'longer_camel_case'],
        ['MixedCamelAnd_snake_case', 'mixed_camel_and_snake_case'],
        ['mixedCamelAnd_snake_case', 'mixed_camel_and_snake_case'],
        ['camelTo_snake_case', 'camel_to_snake_case'],
    ];
}

/**
 * @dataProvider camelToSnakeCaseProvider
 */
public function testCamelToSnakeCase($input, $expected)
{
    $actual = $this->object->camelTo_snake_case($input);
    $this->assertSame($expected, $actual);
}
```

Just a few of notes though:

1. `camelTo_snake_case` does not respect the naming style we use in the code (which is camel case): I'd use `camelToSnakeCase`
2. For the `LongerCamelCase` example you provided you wrote that the expected output is `longerCamelCase`: shouldn't it be `longer_camel_case`? 
3. From your examples I think your new method also accepts in input strings in snake case and pascal case in addidtion to camel case: so it's not just converting from camel case. Correct? If so, what about
   1. supporting kebab case too
   2. renaming the method to something like `toSnakeCase`?

aembler commented:
Agreed strongly with 1 – we don't need to mix cases here.

2: We include a number of laravel helpers (inadvertently at first), but we should check to see whether laravels `snake_case` method, which is accessible in the core already, doesn't already do this exactly as we'd intend.

JohntheFish committed:
Update Text.php

JohntheFish commented:
I ran my tests with the Laravel snake_case and it provided the same output, so no need for this proposed method.

With that in mind I changed the pull to add a comment about Laravel snake_case,  deleted the proposed method and changed the heading of this pull.

I am not fussed either way. Accept the pull in its reduced form to merely add a comment, or reject it as redundant.
